pkg/inject/inject.go

```diff
@@ -0,0 +1,595 @@
+			{
+				Name:  "LINKERD2_PROXY_TLS_POD_IDENTITY",
+				Value: identity.ToDNSName(),
+			},
+			{Name: "LINKERD2_PROXY_CONTROLLER_NAMESPACE", Value: conf.globalConfig.GetLinkerdNamespace()},
+			{Name: "LINKERD2_PROXY_TLS_CONTROLLER_IDENTITY", Value: identity.ToControllerIdentity().ToDNSName()},
+		}
+
+		sidecar.Env = append(sidecar.Env, tlsEnvVars...)
+		sidecar.VolumeMounts = []v1.VolumeMount{
+			{Name: configMapVolume.Name, MountPath: configMapBase, ReadOnly: true},
+			{Name: secretVolume.Name, MountPath: secretBase, ReadOnly: true},
+		}
+
+		if len(conf.podSpec.Volumes) == 0 {
+			patch.addVolumeRoot()
+		}
+		patch.addVolume(configMapVolume)
+		patch.addVolume(secretVolume)
+	}
+
+	patch.addContainer(&sidecar)
+
+	if !conf.globalConfig.GetCniEnabled() {
+		nonRoot := false
+		runAsUser := int64(0)
+		initContainer := &v1.Container{
+			Name:                     k8s.InitContainerName,
+			Image:                    conf.taggedProxyInitImage(),
+			ImagePullPolicy:          v1.PullPolicy(conf.proxyConfig.GetProxyInitImage().GetPullPolicy()),
+			TerminationMessagePolicy: v1.TerminationMessageFallbackToLogsOnError,
+			Args:                     initArgs,
+			SecurityContext: &v1.SecurityContext{
+				Capabilities: &v1.Capabilities{
+					Add: []v1.Capability{v1.Capability("NET_ADMIN")},
+				},
+				Privileged:   &f,
+				RunAsNonRoot: &nonRoot,
+				RunAsUser:    &runAsUser,
+			},
+		}
+		if len(conf.podSpec.InitContainers) == 0 {
+			patch.addInitContainerRoot()
+		}
+		patch.addInitContainer(initContainer)
+	}
+}
+
+// Given a ObjectMeta, update ObjectMeta in place with the new labels and
+// annotations.
+func (conf *ResourceConfig) injectObjectMeta(patch *Patch) {
+	if len(conf.objMeta.Annotations) == 0 {
+		patch.addPodAnnotationsRoot()
+	}
+	patch.addPodAnnotation(k8s.CreatedByAnnotation, k8s.CreatedByAnnotationValue())
+	patch.addPodAnnotation(k8s.ProxyVersionAnnotation, conf.globalConfig.GetVersion())
+
+	if conf.globalConfig.GetIdentityContext() != nil {
+		patch.addPodAnnotation(k8s.IdentityModeAnnotation, k8s.IdentityModeOptional)
+	} else {
+		patch.addPodAnnotation(k8s.IdentityModeAnnotation, k8s.IdentityModeDisabled)
+	}
+
+	if len(conf.objMeta.Labels) == 0 {
+		patch.addPodLabelsRoot()
+	}
+	for k, v := range conf.podLabels {
+		patch.addPodLabel(k, v)
+	}
+}
+
+func (conf *ResourceConfig) taggedProxyImage() string {
+	return fmt.Sprintf("%s:%s",
+		conf.proxyConfig.GetProxyImage().GetImageName(),
+		conf.globalConfig.GetVersion())
+}
+
+func (conf *ResourceConfig) taggedProxyInitImage() string {
+	return fmt.Sprintf("%s:%s",
+		conf.proxyConfig.GetProxyInitImage().GetImageName(),
+		conf.globalConfig.GetVersion())
+}
+
+// shouldInject determines whether or not the given deployment should be
+// injected. A deployment shouldn't be injected if it contains any known sidecars
+// or is on a HostNetwork or the pod is not annotated with "linkerd.io/inject: disabled".
```

Comment:
should this read `or the pod is not annotated with "linkerd.io/inject: disabled"` (remove the `not`?)
Maybe easier read as:
```
// A deployment shouldn't be injected if:
// - it contains any known sidecars; or
// - is on a HostNetwork; or
// - the pod is annotated with "linkerd.io/inject: disabled".
```